functions/src/winloss.ts

```diff
@@ -0,0 +1,122 @@
+  }
+
+  await resetLobby(lobby.ref);
+  await deletePrivatePlayers(lobby.ref, getPrivatePlayerCollection(lobby.ref));
+});
+
+export function findWinner(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+  return db.runTransaction(async (transaction) => {
+    const lobby = await transaction.get(lobbyDocRef);
+    const lobbyData = lobby.data();
+
+    if (lobbyData === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
+    const privatePlayerCollection = getPrivatePlayerCollection(lobbyDocRef);
+    const { players, uids } = lobbyData as Lobby;
+
+    for (let i = 0; i < players.length; i++) {
+      const privatePlayerDocRef = privatePlayerCollection.doc(uids[i]);
+      const privatePlayerDoc = await transaction.get(privatePlayerDocRef);
+      const privatePlayerData = privatePlayerDoc.data();
+
+      if (!privatePlayerData) {
+        throw new Error(`Private player not found for uid ${uids[i]}`);
+      }
+      lobbyData.players[i].role = privatePlayerData.role;
+    }
+
+    let catfishCount = 0;
+    for (let i = 0; i < uids.length; i++) {
+      // check the alive players
+
+      if (lobbyData.players[i].alive == true) {
+        if (lobbyData.players[i].role == "CATFISH") {
+          catfishCount++;
+        }
+      }
+    }
+    if (catfishCount == 0) {
+      lobbyData.winner = "CAT";
+    } else {
+      lobbyData.winner = "CATFISH";
+    }
+
+    transaction.set(lobbyDocRef, lobbyData);
+  });
+}
+
+// when returning to the lobby, make every player alive again and delete the role, winner, and the private player collection
+export function resetLobby(lobbyDocRef: firestore.DocumentReference<Lobby>) {
+  return db.runTransaction(async (transaction) => {
+    const lobby = await transaction.get(lobbyDocRef);
+    const lobbyData = lobby.data();
+
+    if (lobby === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
+
+    if (lobbyData === undefined) {
+      throw new Error("Lobby does not exist!");
+    }
+
+    // loop through each player and set their alive status to true and remove the role property from the player
+    let { players } = lobbyData;
+    players = players.map(({ role, ...rest }) => {
+      return { ...rest, alive: true };
+    });
+
+    transaction.update(lobbyDocRef, { state: "WAIT", players, winner: firestore.FieldValue.delete() });
+  });
+}
+
+export async function deletePrivatePlayers(
+  lobbyDocRef: firestore.DocumentReference<Lobby>,
+  privatePlayerCollectionRef: firestore.CollectionReference<PrivatePlayer>
+) {
+  const lobby = await lobbyDocRef.get();
+  if (lobby.exists === false) {
+    throw new Error("Lobby does not exist!");
+  }
+
+  // loop through each doc in the private player collection via uid and delete them one-by-one
+  const { uids } = lobby.data() as Lobby;
+  for (let i = 0; i <= uids.length; i++) {
+    const playerDoc = privatePlayerCollectionRef.doc(uids[i]);
+    await playerDoc.delete();
```

Comment:
an `await` inside a loop can be really slow maybe consider a batch write https://firebase.google.com/docs/firestore/manage-data/transactions#batched-writes